web/src/lib.rs

```diff
@@ -1,60 +1,291 @@
+        let (on_success_sender, on_success_receiver) = oneshot::channel::<Result<(), ()>>();
+
+        let on_error_closure: Closure<dyn FnMut(_)> = Closure::once(move |_event: Event| {
+            info!("Error occured while trying to get value for key.");
+            // TODO: extract proper error object.
+            on_error_sender.send(Err(())).unwrap();
+        });
+
+        let on_success_closure: Closure<dyn FnMut(_)> = Closure::once(move |_event: Event| {
+            on_success_sender.send(Ok(())).unwrap();
+        });
+
+        request.set_onerror(on_error_closure.as_ref().dyn_ref());
+        request.set_onsuccess(on_success_closure.as_ref().dyn_ref());
+
+        let on_error_receiver_fused = on_error_receiver.fuse();
+        let on_success_receiver_fused = on_success_receiver.fuse();
+
+        pin_mut!(on_error_receiver_fused, on_success_receiver_fused);
+
+        let result: Result<(), ()> = select! {
+            // TODO: pass proper error.
+            _ = on_error_receiver_fused => Err(()),
+            _ = on_success_receiver_fused => Ok(()),
+        };
+
+        result.expect("An error occured while trying to get value for key.");
+        db.close();
+
+        request.set_onerror(None);
+        request.set_onsuccess(None);
+
+        drop(on_error_closure);
+        drop(on_success_closure);
+
+        let js_value = request.result().unwrap();
+        Ok(js_value)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn test_example() {
+        // Nothing here yet, all good.
+    }
+}
+
+#[wasm_bindgen]
+// NOTE: can't use term `Type` here, conflicts with keyword `type` and
+// leads to weird variable names. Therefore "variant" it is.
+pub enum JSStoreVariant {
+    InMemory,
+    IndexedDB
+}
+
 #[wasm_bindgen]
 pub struct JSStore {
-    store: Rc<RefCell<InMemoryStore>>,
+    // TODO: check for thread safety, probably need a replacement for
+    // RefCell, which isn't thread-safe. Also, we probably won't need
+    // dynamic dispatch since we pick the appropriate implementation
+    // via rollup.js (statically).
+    store: Arc<RefCell<dyn Store>>,
 }
 
 #[wasm_bindgen]
 impl JSStore {
 
     #[wasm_bindgen(constructor)]
-    pub fn new(name: &str) -> JSStore {
-        let in_memory_store = InMemoryStore::new(name);
-        let in_memory_store_ref_celled = RefCell::new(in_memory_store);
-        let in_memory_store_reference_counted = Rc::new(in_memory_store_ref_celled);
-        Self { store: in_memory_store_reference_counted }
+    pub fn new(name: &str) -> Promise {
+        // TODO: see TODO comment in struct JSStore, probably no need
+        // for different variants and dynamic dispatch due to rollup.js.
+        // Either get rid of that or make this a parameter and adapt
+        // the native implementation.
+        let variant = JSStoreVariant::IndexedDB;
+
+        console_log::init_with_level(Level::Debug).unwrap();
```

Comment:
No change requested, just curious if the plan to leave the logging statements in for the final project

Reply:
Not really. I'd replace most of it's use-cases with proper error handling and debugging support. I think it's OK to keep it for now, since this is a demo project and we need some way to print stuff to the JS console.